File: Windows Native Runtime/Scene.cpp
#include "Scene.h"
#include <algorithm>
#include "HApplication.h"
#include "ResourceManager.h"
#include "NavigationManager.h"
#include "AnimationManager.h"

void Scene::startScene(SDL_Surface * canvas)
{
	this->canvas = canvas;
	isRunning = true;
	makeSorting();
}

bool Scene::checkSorted(std::list<DrawableGameObject> * index_array)
{
	for (int cnt = 0; cnt < index_array->size( ); cnt++)
	{
		DrawableGameObject base_object;

		struct LoopCounter
		{
			std::list<DrawableGameObject>::iterator cur_obj;
			int cntr;
		};

		for (LoopCounter base_counter = { index_array->begin( ) , 0 }; base_counter.cur_obj != index_array->end(); base_counter.cur_obj++, base_counter.cntr++)
		{
			if (base_counter.cntr == cnt)
			{
				base_object = *(base_counter.cur_obj);
				break;
			}
		}

		if (cnt + 1 < index_array->size( ))
		{
			DrawableGameObject cmp_object;

			for (LoopCounter base_counter = { index_array->begin(), 0 }; base_counter.cur_obj != index_array->end(); base_counter.cur_obj++, base_counter.cntr++)
			{
				if (base_counter.cntr == cnt + 1)
				{
					cmp_object = *(base_counter.cur_obj);
					break;
				}
			}

			if (base_object.depth < cmp_object.depth)
			{
				return false;
			}
		}
	}

	return true;
}

void Scene::sortElements(std::list<DrawableGameObject> * index_array)
{
	for (int cnt = 0; cnt < index_array->size( ); cnt++)
	{
		DrawableGameObject base_object;
		std::list<DrawableGameObject>::iterator base_object_iter;

		struct LoopCounter
		{
			std::list<DrawableGameObject>::iterator cur_obj;
			int cntr;
		};

		for (LoopCounter base_counter = { index_array->begin(), 0 }; base_counter.cur_obj != index_array->end(); base_counter.cur_obj++, base_counter.cntr++)
		{
			if (base_counter.cntr == cnt)
			{
				base_object = *(base_counter.cur_obj);
				base_object_iter = base_counter.cur_obj;
				break;
			}
		}

		for (int c = cnt + 1; c < index_array->size( ); c++)
		{
			DrawableGameObject cmp_object;
			std::list<DrawableGameObject>::iterator cmp_object_iter;

			for (LoopCounter base_counter = { index_array->begin(), 0 }; base_counter.cur_obj != index_array->end(); base_counter.cur_obj++, base_counter.cntr++)
			{
				if (base_counter.cntr == c)
				{
					cmp_object = *(base_counter.cur_obj);
					cmp_object_iter = base_counter.cur_obj;
					break;
				}
			}

			if (base_object.depth < cmp_object.depth)
			{
				DrawableGameObject cp = base_object;

				*base_object_iter = cmp_object;
				*cmp_object_iter = cp;

				break;
			}
		}
	}
}

void Scene::makeSorting()
{
	sortedArray = std::list<DrawableGameObject>( );

	for (int cnt = 0; cnt < object_array.size( ); cnt++)
	{
		GameObject_Scene_ptr obj = NULL;

		struct LoopCounter
		{
			std::list<GameObject_Scene_ptr>::iterator cur_obj;
			int cntr;
		};

		for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++,base_counter.cntr++)
		{
			if (base_counter.cntr == cnt)
			{
				obj = *(base_counter.cur_obj);
				break;
			}
		}

		DrawableGameObject dobj;

		dobj.depth = obj->depth;
		dobj.index = cnt;

		sortedArray.push_back(dobj);
	}

	while (!checkSorted(&sortedArray))
	{
		sortElements(&sortedArray);
	}
}

void Scene::updateScene()
{
	bool hasDeleted = false;
	
	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

re:
	for (int cnt = 0; cnt < object_array.size( ); cnt++)
	{
		GameObject_Scene_ptr obj = NULL;

		LoopCounter base_counter = { object_array.begin(), 0 };

		for (; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++,base_counter.cntr++)
		{
			if (base_counter.cntr == cnt)
			{
				obj = *(base_counter.cur_obj);
				break;
			}
		}

		if (obj->isDestroyed)
		{
			object_array.erase(base_counter.cur_obj);
			hasDeleted = true;
			goto re;
		}
	}

	if (hasDeleted) makeSorting();

	for (int cnt = 0; cnt < object_array.size( ); cnt++)
	{
		GameObject_Scene_ptr obj = NULL;

		for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++, base_counter.cntr++)
		{
			if (base_counter.cntr == cnt)
			{
				obj = *(base_counter.cur_obj);
				break;
			}
		}

		if (!obj->obj_instance._static)
		{
			if (obj->obj_instance.physics)
			{
				if (obj->obj_instance.rigidbody)
				{
					obj->pos_y += gravity;
				}

				for (int cnt0 = 0; cnt0 < object_array.size( ); cnt0++)
				{
					if (cnt0 == cnt)
					{
						continue;
					}
					else
					{
						GameObject_Scene_ptr cmp_obj = NULL;

						for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++, base_counter.cntr++)
						{
							if (base_counter.cntr == cnt0)
							{
								cmp_obj = *(base_counter.cur_obj);
								break;
							}
						}

						if (obj->obj_instance.img.baseImage != NULL && cmp_obj->obj_instance.img.baseImage != NULL && obj->obj_instance.physics && cmp_obj->obj_instance.physics && obj->obj_instance.collider && cmp_obj->obj_instance.collider)
						{
							if (obj->pos_x + obj->obj_instance.img.Width > cmp_obj->pos_x && obj->pos_x < cmp_obj->pos_x + cmp_obj->obj_instance.img.Width && obj->pos_y + obj->obj_instance.img.Height > cmp_obj->pos_y && cmp_obj->pos_y < cmp_obj->pos_y + cmp_obj->obj_instance.img.Height)
							{
								std::list<OnCollisionHandler> col_handler = HApplication::getCollisionHandlers();

								for (std::list<OnCollisionHandler>::iterator cur_handler = col_handler.begin(); cur_handler != col_handler.end(); cur_handler++)
								{
									(*cur_handler)(obj,cmp_obj);
								}
							}
						}
					}
				}
			}

		}

		SDL_Event e;

		while (SDL_PollEvent(&e))
		{
			if (e.type == SDL_KEYDOWN)
			{
				std::list<OnKeyDown> keydown_handlers = HApplication::getKeyboardHandlers();

				for (std::list<OnKeyDown>::iterator cur_handler = keydown_handlers.begin(); cur_handler != keydown_handlers.end(); cur_handler++)
				{
					(*cur_handler)(e.key.keysym.sym);
				}
			}
			else if (e.type == SDL_QUIT)
			{
				HApplication::Exit();
			}
			else if (e.type == SDL_MOUSEBUTTONDOWN)
			{
				if (e.button.button == SDL_BUTTON_LEFT || e.button.button == SDL_BUTTON_RIGHT)
				{
					std::list<OnMouseDown> mousedown_handers = HApplication::getMouseDownHandlers();

					for (std::list<OnMouseDown>::iterator cur_handler = mousedown_handers.begin(); cur_handler != mousedown_handers.end();cur_handler++)
					{
						(*cur_handler)(e.button.button);
					}
				}
			}
			else if (e.type == SDL_MOUSEMOTION)
			{
				std::list<OnMouseMove> mousemove_handlers = HApplication::getMouseMoveHandlers();

				for (std::list<OnMouseMove>::iterator cur_handler = mousemove_handlers.begin(); cur_handler != mousemove_handlers.end(); cur_handler++)
				{
					(*cur_handler)(e.button.x, e.button.y);
				}
			}
		}

		if (!obj->scriptsEmpty())
		{
			obj->processScripts();
		}
	}

	NavigationManager::updateNavigation();
	AnimationManager::updateAnimation();

	drawScene();
	SDL_Delay(speed);
}

void Scene::drawScene()
{
	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

	struct LoopCounter_DrawableGameObject
	{
		std::list<DrawableGameObject>::iterator cur_obj;
		int cntr;
	};

	SDL_FillRect(this->canvas, GetRectP(0, 0, HApplication::getSize().w, HApplication::getSize().h),SDL_MapRGBA(canvas->format,R,G,B,A));

	for (int cntr = 0; cntr < sortedArray.size( ); cntr++)
	{
		DrawableGameObject obj;

		for (LoopCounter_DrawableGameObject base_counter = { sortedArray.begin(), 0 }; base_counter.cur_obj != sortedArray.end(); base_counter.cur_obj++, base_counter.cntr++)
		{
			if (cntr == base_counter.cntr)
			{
				obj = *(base_counter.cur_obj);
				break;
			}
		}

		int cnt = obj.index;

		GameObject_Scene_ptr base_obj = NULL;

		for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++, base_counter.cntr++)
		{
			if (cnt == base_counter.cntr)
			{
				base_obj = *(base_counter.cur_obj);
				break;
			}
		}

		if (base_obj->obj_instance.img.baseImage != NULL)
		{
			SDL_BlitSurface(base_obj->obj_instance.img.baseImage, NULL, canvas,GetRectP(base_obj->pos_x,base_obj->pos_y,base_obj->obj_instance.img.Width,base_obj->obj_instance.img.Height));
		}
		else if (base_obj->obj_instance.text != "")
		{
			TTF_Font * baseFont = TTF_OpenFont(ResourceManager::getResource(base_obj->obj_instance.font_name).c_str(), base_obj->obj_instance.font_size);
			SDL_Surface * baseSurface = TTF_RenderText_Solid(baseFont, base_obj->obj_instance.text.c_str(), base_obj->obj_instance.color);
			
			SDL_BlitSurface(baseSurface, NULL, canvas, GetRectP(base_obj->pos_x,base_obj->pos_y,0,0));
		}
	}

	SDL_UpdateWindowSurface(HApplication::getWindowHandle());
}

void Scene::endScene()
{
	this->isRunning = false;
}

bool Scene::loadGameObject(GameObject_Scene_ptr gameObject)
{
	if (gameObject->instance_name == "")
	{
		return false;
	}

	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

	for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++)
	{
		GameObject_Scene_ptr gameObj = *(base_counter.cur_obj);

		if (gameObj->instance_name == gameObject->instance_name)
		{
			return false;
		}
	}

	object_array.push_back(gameObject);
	makeSorting();

	return true;
}

void Scene::destroyGameObject(std::string instance_name)
{
	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

	for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++)
	{
		GameObject_Scene_ptr gameObj = *(base_counter.cur_obj);

		if (gameObj->instance_name == instance_name)
		{
			gameObj->isDestroyed = true;
			return;
		}
	}
}

GameObject_Scene_ptr Scene::findGameObject(std::string name)
{
	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

	for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++)
	{
		GameObject_Scene_ptr gameObj = *(base_counter.cur_obj);

		if (gameObj->instance_name == name)
		{
			return gameObj;
		}
	}

	return NULL;
}

std::list<GameObject_Scene_ptr> Scene::findGameObject(int tag)
{
	std::list<GameObject_Scene_ptr> ret_list;

	struct LoopCounter
	{
		std::list<GameObject_Scene_ptr>::iterator cur_obj;
		int cntr;
	};

	for (LoopCounter base_counter = { object_array.begin(), 0 }; base_counter.cur_obj != object_array.end(); base_counter.cur_obj++)
	{
		GameObject_Scene_ptr gameObj = *(base_counter.cur_obj);

		if (gameObj->obj_instance.tag == tag)
		{
			ret_list.push_back(gameObj); 
		}
	}

	return ret_list;
}

File: Java Desktop Runtime/jruntime/Scene.java
package jruntime;

import jruntime.CollisionHandler;
import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.io.*;
import net.coobird.thumbnailator.Thumbnails;
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author Riju
 */
public class Scene implements Runnable , Cloneable
{
    ArrayList object_array;
    JPanel canvas;        
    public int A = 0,R = 0,G = 0,B = 0;
    public int speed = 1;
    public int gravity = 0;
    public String name = "";
    private boolean isRunning = false;
    public ArrayList collision_handlers;	
    DrawableGameObject[] sortedArray;
    
    class DrawableGameObject
    {
        int depth;
        int index;
    }
    
    public Scene( )
    {
        object_array = new ArrayList( );
        collision_handlers = new ArrayList( );
        sortedArray = new DrawableGameObject[0];
    }
    
    public void startScene(JPanel canvas)
    {
         this.canvas = canvas;
         isRunning = true;
         makeSorting( );
         
	 Thread t = new Thread(this);
         t.start();
    }
    
    @Override
    public void run( )
    {
        while(isRunning)
        {
            updateScene( );
            
            try
            {
                Thread.sleep(speed);
            }
            catch(InterruptedException ax)
            {
                
            }
        }
    }
		
		private void updateScene()
		{
                   boolean hasDeleted = false,hasDel = false; 
                    
                   do
                   {
                       hasDeleted = false;
                       
                       for(int cnt = 0;cnt < object_array.size();cnt++)
                       {
                           if (((GameObject_Scene) object_array.get(cnt)).isDestroyed)
                           {
                               object_array.remove(cnt);
                               hasDeleted = true;
                               hasDel = true;
                               break;
                           }
                       }
                       
                   } while(hasDeleted);
                   
                   if (hasDel) makeSorting( );
                   
                   
		   for (int cnt = 0;cnt < object_array.size();cnt++)
                   {
			     if (((GameObject_Scene) object_array.get(cnt)).obj_instance._static == false)
				 {
					if (((GameObject_Scene) object_array.get(cnt)).obj_instance.physics)
					{
						if (((GameObject_Scene) object_array.get(cnt)).obj_instance.rigidbody)
						{
							((GameObject_Scene) object_array.get(cnt)).Translate(0,gravity);
						}
						
						for(int cnt0 = 0;cnt0 < object_array.size( );cnt0++)
						{
						      if (cnt0 == cnt)
							  {
							    continue;
							  }
							  else
						      {
							    if (((GameObject_Scene) object_array.get(cnt)).obj_instance.img != null && ((GameObject_Scene) object_array.get(cnt0)).obj_instance.img != null && ((GameObject_Scene) object_array.get(cnt)).obj_instance.physics && ((GameObject_Scene) object_array.get(cnt0)).obj_instance.physics && ((GameObject_Scene) object_array.get(cnt)).obj_instance.collider && ((GameObject_Scene) object_array.get(cnt0)).obj_instance.collider)
								{
									if (((GameObject_Scene) object_array.get(cnt)).pos_x + ((GameObject_Scene) object_array.get(cnt)).obj_instance.img.getWidth( ) > ((GameObject_Scene) object_array.get(cnt0)).pos_x && ((GameObject_Scene) object_array.get(cnt)).pos_x < ((GameObject_Scene) object_array.get(cnt0)).pos_x + ((GameObject_Scene) object_array.get(cnt0)).obj_instance.img.getWidth( ) && ((GameObject_Scene) object_array.get(cnt)).pos_y + ((GameObject_Scene) object_array.get(cnt)).obj_instance.img.getHeight( ) > ((GameObject_Scene) object_array.get(cnt0)).pos_y && ((GameObject_Scene) object_array.get(cnt)).pos_y < ((GameObject_Scene) object_array.get(cnt0)).pos_y + ((GameObject_Scene) object_array.get(cnt0)).obj_instance.img.getHeight( ) && ((GameObject_Scene) object_array.get(cnt)).depth == ((GameObject_Scene) object_array.get(cnt0)).depth)
									{
                                                                            
                                                                            for(int c = 0;c < collision_handlers.size( );c++)
                                                                            {
                                                                                ((CollisionHandler) collision_handlers.get(c)).onCollision((GameObject_Scene) object_array.get(cnt),(GameObject_Scene) object_array.get(cnt0));
                                                                            }
									}
								}
							  }
						}
					}
                                 }
                       
                             ((GameObject_Scene) object_array.get(cnt)).processScripts();
		   }	
                   
                   NavigationManager.updateNavigation();
                   AnimationManager.updateAnimation();
                   NetworkManager.updateNetwork();
		}
		
                 public boolean checkSorted(DrawableGameObject[] index_array)
        {
            for (int cnt = 0; cnt < index_array.length; cnt++)
            {
                if (cnt + 1 < index_array.length)
                {
                    if (index_array[cnt].depth < index_array[cnt + 1].depth)
                    {
                        return false;
                    }
                }
            }

            return true;
        }

        public void sortElements(DrawableGameObject[] index_array)
        {
            for (int cnt = 0; cnt < index_array.length; cnt++)
            {
                for (int c = cnt + 1; c < index_array.length; c++)
                {
                    if (index_array[cnt].depth < index_array[c].depth)
                    {
                        DrawableGameObject cp = index_array[cnt];

                        index_array[cnt] = index_array[c];

                        index_array[c] = cp;

                        break;
                    }
                }
            }
        }

        private void makeSorting()
        {
            sortedArray = new DrawableGameObject[object_array.size()];

            for (int cnt = 0; cnt < object_array.size( ); cnt++)
            {
                sortedArray[cnt] = new DrawableGameObject( );
                sortedArray[cnt].depth = ((GameObject_Scene) object_array.get(cnt)).depth;
                sortedArray[cnt].index = cnt;
            }

            while (!checkSorted(sortedArray))
            {
                sortElements(sortedArray);
            }
        }
           
         public ArrayList<GameObject_Scene> getAllGameObjects( ){ return object_array; }
        
		public void drawScene(Graphics g)
		{
                    g.setColor(new Color(R,G,B,A));
		    g.fillRect(0,0,canvas.getWidth(),canvas.getHeight());
			
		    for(int cnt = 0;cnt < object_array.size();cnt++)
			{
               if (((GameObject_Scene) object_array.get(cnt)).obj_instance.img != null)
			   {
                             try
                             {
                                 g.drawImage(Thumbnails.of(((GameObject_Scene) object_array.get(cnt)).obj_instance.img).forceSize(((GameObject_Scene) object_array.get(cnt)).obj_instance.img.getWidth() + (int) ((GameObject_Scene) object_array.get(cnt)).GetScale(), ((GameObject_Scene) object_array.get(cnt)).obj_instance.img.getHeight() + (int) ((GameObject_Scene) object_array.get(cnt)).GetScale()).asBufferedImage(),((GameObject_Scene) object_array.get(cnt)).pos_x,((GameObject_Scene) object_array.get(cnt)).pos_y,null);
                             }
                             catch(IOException ex) { }
			   }			   
			   else if (!((GameObject_Scene) object_array.get(cnt)).obj_instance.text.equals(""))
			   {
                               g.setFont(new Font(((GameObject_Scene) object_array.get(cnt)).obj_instance.font_name,Font.PLAIN,((GameObject_Scene) object_array.get(cnt)).obj_instance.font_size));
                               g.setColor(((GameObject_Scene)  object_array.get(cnt)).obj_instance.color);
                  g.drawString(((GameObject_Scene) object_array.get(cnt)).obj_instance.text,((GameObject_Scene) object_array.get(cnt)).pos_x,((GameObject_Scene) object_array.get(cnt)).pos_y);				  
			   }
			}
		}
		
		public void endScene( )
		{
		 isRunning = false;	
		}
		
		public boolean loadGameObject(GameObject_Scene gameObject)
		{
                   if (gameObject.instance_name.equals(""))
                   {
                       return false;
                   } 
                   
                   for(int cnt = 0;cnt < object_array.size();cnt++)
                   {
                       if (((GameObject_Scene) object_array.get(cnt)).instance_name.equals(gameObject.instance_name))
                       {
                           return false;
                       }
                   }
                   
                   object_array.add(gameObject);
                   makeSorting( );
                   
                   return true;
		}
		
                public void destroyGameObject(String instance_name )
                { 
                    for(int cnt = 0;cnt < object_array.size( );cnt++)
                    {
                        if (((GameObject_Scene) object_array.get(cnt)).instance_name.equals(instance_name))
                        {
                            ((GameObject_Scene) object_array.get(cnt)).isDestroyed = true;
                            return;
                        }
                    }
                }
                
		public GameObject_Scene findGameObject(String name)
		{
			for(int cnt = 0;cnt < object_array.size( );cnt++)
			{
                            if ( ((GameObject_Scene) object_array.get(cnt)).instance_name.equals(name))
			   {
					return (GameObject_Scene) object_array.get(cnt);
			   }			   
			}
			
			return null;
		}
                
                 public  ArrayList<GameObject_Scene> findGameObject(int tag)
                 {
                    ArrayList<GameObject_Scene> ret_list = new ArrayList<GameObject_Scene>( );
               
                    for(int cnt = 0;cnt < object_array.size( );cnt++)
                    {
                          if (((GameObject_Scene) object_array.get(cnt)).obj_instance.tag == tag)
			  {
                              ret_list.add((GameObject_Scene) object_array.get(cnt));
			  }
                    }		  
		  
                    return ret_list;
                 }
                
                public void registerCollisionHandler(CollisionHandler handle)
                {
                    collision_handlers.add(handle);
                }
                
                @Override
                protected Object clone( ) throws CloneNotSupportedException
                {
                    return super.clone();
                }
	}
